python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -1247,29 +1256,20 @@ def diff(self, periods=1, axis=0):
         if not axis == 0:
             raise NotImplementedError("Only axis=0 is supported.")
 
-        # grouped values
-        value_columns = self.grouping.values
-        _, (data, index), _ = self._groupby.groups(
-            cudf.core.frame.Frame(value_columns._data)
-        )
-        grouped = self.obj.__class__._from_data(data, index)
-        grouped = self._mimic_pandas_order(grouped)._copy_type_metadata(
-            value_columns
+        values = self.obj.__class__._from_data(
+            self.grouping.values._data, self.obj.index
         )
-
-        result = grouped - self.shift(periods=periods)
-        return result._copy_type_metadata(value_columns)
+        return values - self.shift(periods=periods)
 
     def _scan_fill(self, method: str, limit: int) -> DataFrameOrSeries:
         """Internal implementation for `ffill` and `bfill`"""
-        value_columns = self.grouping.values
-        result = self.obj.__class__._from_data(
-            self._groupby.replace_nulls(
-                cudf.core.frame.Frame(value_columns._data), method
-            )
+        values = self.grouping.values
+        result = self.obj._from_columns(
+            self._groupby.replace_nulls([*values._columns], method),
```

Comment:
I see a list comprehension/unpacking like `[*stuff._columns]` used all over in this PR. Is `stuff._columns` a tuple? Can you pass that tuple directly, without needing to convert it to a list? It makes sense if you need to combine two iterables, like `[*first, *second]`, but this is happening for single iterables. This applies in many places.
```suggestion
            self._groupby.replace_nulls(values._columns, method),
```

Reply:
Cython will complain. That is one of the tradeoffs in strongly typing Cython functions as accepting a list. We could revisit that decision (I know @shwina felt strongly that it would be very helpful for debugging purposes, especially for devs who aren't Cython experts) but I'd say that is out of scope for this PR.

Reply:
Oh, and yes, [`Frame._columns` is a tuple](https://github.com/rapidsai/cudf/blob/branch-22.04/python/cudf/cudf/core/frame.py#L167).

Reply:
> Does Cython complain about type mismatch, or does it cast?

It complains. @vyasr and I chatted about this: on the one hand we want to use some typing to ensure type safety, we settled on list originally but I think this is a use case that drive us towards using tuple as the type.

Reply:
Well, I wouldn't necessarily say that this is a reason to use tuples since it's not immediately clear to me that a tuple is the correct return type for `Frame._columns` a priori. I think I have seen its usage in ways that sometimes support list and sometimes support tuple. We should probably assess that a little bit more carefully before rushing to any conclusions here.

Reply:
This is not a dealbreaker for this PR, I’d merge it as-is. I think my argument to @shwina for making the columns into a tuple was to indicate immutability. However, it would be fine with me to revisit that decision for the sake of convenience and reducing the need for workarounds.

Reply:
Are you just talking about type annotations when you say "making columns into a tuple"? AFAIK `ColumnAccessor.columns` (which is called by `Frame._columns`) has _always_ returned a tuple. But yes, the annotation was wrong for a long time, and it could still be improved. Modifying the behavior is a slightly larger change though since a lot of code probably relies on accessing the columns through this property.

Reply:
Yes, that must be what I mean. I only remembered that the type annotation didn’t match the behavior.